backend/amd64/proc.ml

```diff
@@ -102,21 +102,32 @@ let num_register_classes = 2
 let register_class r =
   match r.typ with
   | Val | Int | Addr -> 0
-  | Float | Vec128 -> 1
-
-let register_class_tag c =
-  match c with
-  | 0 -> "i"
-  | 1 -> "f"
-  | c -> Misc.fatal_errorf "Unspecified register class %d" c
+  | Float -> 1
+  | Vec128 ->
+    if not !simd_regalloc_support then
+    Misc.fatal_error "SIMD register allocation is not enabled.";
+    1
```

Comment:
It's not clear to me that `register_class` can't be called on the result of `phys_reg` triggering the assert, but we will find out. 

It is unfortunate that `phys_reg` will readily return a `Reg.t` with `typ=vec128` even when simd_regalloc_support flag is off. I'd rather never create any `Reg.t` with that type when the flag is off. Is the only reason to creat them is to preallocate `hard_vec128_reg`?

Related to it, the distinction between the `precolored_regs` and `all_phys_regs` is not clear to me. They depend on the same compile-time flag. In fact, `precolored_regs` depends on `num_available_registers` that depens on `fp` with is a configure-time flag, so should be known statically. Concretely, `precolored_regs` can return a set not an array, using `Reg.set_of_array (all_phys_reg ())` and then `Reg.Set.remove rbp` from it if `fp` is true.

Reply:
Made phys_reg check the flag, which means `destroyed_at_c_call` now uses `hard_vec128_reg` directly. I think the result is slightly better.

Reply:
Refactored again